# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
print(lambda num: num +1)
print((lambda num: num +1)(3))

<function <lambda> at 0x0000019E94D64D30>
4


#### Saving to a Variable

In [2]:
adone = lambda num: num + 1
print(adone(4))

5


#### Multiple Inputs

In [9]:
find_vol = lambda length,width,height: length * width * height
print(find_vol(2,2,3))

12


#### Passing a Lambda into a Function

In [5]:
def multiply(function, num):
    return function(num)
print(multiply(lambda num: num * num,4))

16


#### Returning a Lambda from a Function

In [7]:
def set_exponent(num):
    return lambda l_num: l_num ** num
square = set_exponent(2)
print(square(4))
cubed = set_exponent(3)
print(cubed(5))

16
125


#### If Statements within Lambdas

In [8]:
is_even = lambda num: f"{num} is even" if num % 2 == 0 else f"{num} is odd"
print(is_even(5))

5 is odd


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [10]:
f = lambda num: num ** 3
print(f(3))

27


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [12]:
my_list = [12,45,2,34]
def double_num(alist):
    double = []
    for num in alist:
        double.append(num*2)
        print(double_nums(my_list))
        def mao_double(num):
          return num * 2
    list(map(map_double,my_list))

#### Using Lambda's with Map

In [14]:
print(list(map(lambda num: num * 2,my_list)))

[24, 90, 4, 68]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [15]:
#[2,6,99,2,45,56]
my_list = [2,6,99,2,45,56]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [65]:
def only_odds(alist):
    odd_nums = []
    for num in alist:
        if num % 2 != 0:
            odd_nums.append(num)
    return odd_nums

new_list = [1,12,54,41,23456,23]

print(only_odds(new_list))


# True if (condition) else False

#return True if num %2 != 0 else False 

#return num %2 != 0

def filter_odds(num):
    return num %2 != 0 

print(list(filter(filter_odds, new_list)))

            

[1, 41, 23]
[1, 41, 23]


#### Using Lambda's with Filter()

In [66]:
print(list(filter(lambda num: num %2 != 0 , new_list)))

print(2 % 2 != 0)

[1, 41, 23]
False


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [67]:
from functools import reduce

new_list = [1,2,3,4,5,6]

def add_nums(alist):
    total = 0 
    for num in alist:
        total += num
    return total

print(add_nums(new_list))

def reduce_add(num1, num2):
    print(num1, num2)
    return num1 * num2

print(reduce(reduce_add, new_list))

21
1 2
2 3
6 4
24 5
120 6
720


#### Using Lambda's with Reduce()

In [68]:
reduce(lambda num1, num2: num1 + num2, new_list)

21

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [69]:
reduce(lambda num1, num2: num1 * num2, new_list)

720

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [70]:
def add_nums(num):
    # 5 + 4 + 3 + 2 + 1 = 15
    #base case
    if num <=1:
        print(f'add_nums({num}) = {num}')
        return num
    else:
        #recursive case
        print(f'add_nums({num}) = {num} + add_nums({num-1})')
        return num * add_nums(num - 1)
    
print(add_nums(5))

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1
120


#### Writing a Factorial Function

In [71]:
def factorial(num):
    if num <= 1:
        return num
    else:
        return num * factorial(num-1)
    
print(factorial(5))

120


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [62]:
def sqr_gen(num):
    while num > 0:
        yield num ** 2
        num -= 1

for i in sqr_gen(10):
    print(i)


100
81
64
49
36
25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [61]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
trim_list = [x.strip() for x in places] 
modified_list = [x for x in trim_list if x]
print("modified_list is:"+ str(modified_list))



modified_list is:['Argentina', 'San Diego', 'Boston', 'New York']


author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
author.sort(key = lambda x : (x.split()[-1].lower()))
print(author)



### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [72]:
#F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
newplaces = list(map(lambda c: (c[0], (9/5)*c[1]+32,places)))


TypeError: map() must have at least two arguments.

w### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`